server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java

```diff
@@ -101,47 +106,64 @@ public int getNumberOfInFlightFetch() {
         return count;
     }
 
+    @Override
     public void applyStartedShards(final RoutingAllocation allocation, final List<ShardRouting> startedShards) {
         for (ShardRouting startedShard : startedShards) {
             Releasables.close(asyncFetchStarted.remove(startedShard.shardId()));
             Releasables.close(asyncFetchStore.remove(startedShard.shardId()));
         }
     }
 
+    @Override
     public void applyFailedShards(final RoutingAllocation allocation, final List<FailedShard> failedShards) {
         for (FailedShard failedShard : failedShards) {
             Releasables.close(asyncFetchStarted.remove(failedShard.getRoutingEntry().shardId()));
             Releasables.close(asyncFetchStore.remove(failedShard.getRoutingEntry().shardId()));
         }
     }
 
-    public void allocateUnassigned(final RoutingAllocation allocation) {
+    @Override
+    public void beforeAllocation(final RoutingAllocation allocation) {
         assert primaryShardAllocator != null;
         assert replicaShardAllocator != null;
         ensureAsyncFetchStorePrimaryRecency(allocation);
-        innerAllocatedUnassigned(allocation, primaryShardAllocator, replicaShardAllocator);
     }
 
-    // allow for testing infra to change shard allocators implementation
-    protected static void innerAllocatedUnassigned(RoutingAllocation allocation,
-                                                   PrimaryShardAllocator primaryShardAllocator,
-                                                   ReplicaShardAllocator replicaShardAllocator) {
-        RoutingNodes.UnassignedShards unassigned = allocation.routingNodes().unassigned();
-        unassigned.sort(PriorityComparator.getAllocationComparator(allocation)); // sort for priority ordering
-
-        primaryShardAllocator.allocateUnassigned(allocation);
+    @Override
+    public void afterPrimariesBeforeReplicas(RoutingAllocation allocation) {
+        assert replicaShardAllocator != null;
```

Comment:
I find this assert unnecessary since it is a final field and it is initialized in constructor?

Reply:
It is `null` in tests that use a subclass, and I found it all-too-easy to trip over this.

I don't really like how the tests override this class's behaviour, but don't want to start working on that kind of refactoring here.

Reply:
Ah yes, I overlooked that, thanks.